In [1]:
#Подключим нужные библиотеки
import numpy as np
import pandas as pd

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import RandomizedSearchCV

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from catboost import CatBoostClassifier
from sklearn.metrics import classification_report, accuracy_score

from scipy.stats import randint, uniform

Так как ранее данные уже были проанализированны и очишенны от выбросов и неэффективгых признаков сразу приступаем к построению моделей.

In [2]:
#Загрузи данные
SI_df = pd.read_csv('./data/SI.csv')
display(SI_df)

Unnamed: 0         SI  MaxEStateIndex  MinEStateIndex       qed  \
0             0  28.125000        5.094096        0.387225  0.417362   
1             1   7.000000        3.961417        0.533868  0.462473   
2             2   0.720000        2.627117        0.543231  0.260923   
3             3  63.235294        5.097360        0.390603  0.377846   
4             4   1.300000        5.150510        0.270476  0.429038   
..          ...        ...             ...             ...       ...   
885         996   1.129017       12.934891       -0.476142  0.382752   
886         997   1.062484       13.635345       -0.699355  0.369425   
887         998   1.096761       13.991690       -0.650790  0.284923   
888         999   1.031272       13.830180       -1.408652  0.381559   
889        1000   1.000000       13.380863       -0.447978  0.452565   

           SPS  HeavyAtomMolWt  MaxPartialCharge  MinPartialCharge  \
0    42.928571         340.300          0.038844         -0.293526   
1    45.214286         340.300          0.012887         -0.313407   
2    42.187500         388.344          0.094802         -0.325573   
3    41.862069         352.311          0.038844         -0.293526   
4    36.514286         424.377          0.062897         -0.257239   
..         ...             ...               ...               ...   
885  49.133333         380.270          0.317890         -0.468587   
886  44.542857         446.309          0.327562         -0.467493   
887  41.973684         502.398          0.327887         -0.467485   
888  39.000000         480.299          0.312509         -0.468755   
889  48.580645         388.293          0.311311         -0.468587   

     MaxAbsPartialCharge  ...  fr_morpholine  fr_nitro  fr_para_hydroxylation  \
0               0.293526  ...              0         0                      0   
1               0.313407  ...              0         0                      0   
2               0.325573  ...              0         0                      0   
3               0.293526  ...              0         0                      0   
4               0.257239  ...              0         0                      0   
..                   ...  ...            ...       ...                    ...   
885             0.468587  ...              0         0                      0   
886             0.467493  ...              0         0                      0   
887             0.467485  ...              0         0                      0   
888             0.468755  ...              0         0                      0   
889             0.468587  ...              0         0                      0   

     fr_piperdine  fr_priamide  fr_quatN  fr_sulfide  fr_sulfonamd  \
0               0            0         0           0             0   
1               0            0         0           0             0   
2               0            0         2           0             0   
3               0            0         0           0             0   
4               0            0         0           0             0   
..            ...          ...       ...         ...           ...   
885             0            0         0           0             0   
886             0            0         0           0             0   
887             0            0         0           1             0   
888             0            0         0           0             0   
889             0            0         0           0             0   

     fr_thiazole  fr_unbrch_alkane  
0              0                 3  
1              0                 3  
2              0                 3  
3              0                 4  
4              0                 0  
..           ...               ...  
885            0                 0  
886            0                 0  
887            0                 0  
888            0                 0  
889            0                 0  

[890 rows x 122 columns]

Преобразим целеавой признак в True и False для выполнения классификации.

Значение True установим если значение SI > медианного значения иначе установим False.

Так как медианное значение это значение центрального элемента выборки, то имеем сбалансированный датасет для обучения.

In [4]:
#Преобразим целевой признак
target = SI_df['SI'] >= SI_df['SI'].median()

In [5]:
#Выведем признаки
features = SI_df.drop(['SI', 'Unnamed: 0'], axis=1)

In [6]:
#Выполним стандартизацию данных
scaler = StandardScaler()
features = scaler.fit_transform(features)

In [7]:
#Разделение на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)

<h2>Модель логистической регрессии</h2>

In [8]:
model = LogisticRegression(
    max_iter=200
)
model.fit(X_train, y_train)

# Предсказание
y_pred = model.predict(X_test)
print('Логистическая регрессия:')
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(classification_report(y_test, y_pred))

Логистическая регрессия:
Accuracy: 0.6741573033707865
              precision    recall  f1-score   support

       False       0.69      0.64      0.66        89
        True       0.66      0.71      0.68        89

    accuracy                           0.67       178
   macro avg       0.67      0.67      0.67       178
weighted avg       0.67      0.67      0.67       178



<h2>Модель DecisionTree</h2>

In [18]:
model = DecisionTreeClassifier(
    max_depth=7,
    min_samples_split=4,
    min_samples_leaf=1,
    random_state=42
)
model.fit(X_train, y_train)

# Предсказание
y_pred = model.predict(X_test)
print('Дерево решений:')
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(classification_report(y_test, y_pred))

Дерево решений:
Accuracy: 0.6741573033707865
              precision    recall  f1-score   support

       False       0.64      0.81      0.71        89
        True       0.74      0.54      0.62        89

    accuracy                           0.67       178
   macro avg       0.69      0.67      0.67       178
weighted avg       0.69      0.67      0.67       178



<h2>Модель Random Forest</h2>

In [31]:
#Обучение модели
model = RandomForestClassifier(
    n_estimators=65,
    max_depth=8,
    max_features='log2',
    random_state=42
)
model.fit(X_train, y_train)

# Предсказание
y_pred = model.predict(X_test)
print('Ансамбль деревьев:')
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(classification_report(y_test, y_pred))

Ансамбль деревьев:
Accuracy: 0.702247191011236
              precision    recall  f1-score   support

       False       0.70      0.72      0.71        89
        True       0.71      0.69      0.70        89

    accuracy                           0.70       178
   macro avg       0.70      0.70      0.70       178
weighted avg       0.70      0.70      0.70       178



<h2>Модель SVС</h2>

In [43]:
# Обучение модели
model = SVC(
    kernel='rbf',
    class_weight='balanced',
    C=250
)
model.fit(X_train, y_train)

# Предсказание
y_pred = model.predict(X_test)
print('Классивикация по опорным векторам:')
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(classification_report(y_test, y_pred))

Классивикация по опорным векторам:
Accuracy: 0.6685393258426966
              precision    recall  f1-score   support

       False       0.69      0.62      0.65        89
        True       0.65      0.72      0.68        89

    accuracy                           0.67       178
   macro avg       0.67      0.67      0.67       178
weighted avg       0.67      0.67      0.67       178



<h2>Модель KNN</h2>

In [52]:
# Обучение модели
model = KNeighborsClassifier(
    n_neighbors=5,
    #weights='distance'
)
model.fit(X_train, y_train)

# Предсказание
y_pred = model.predict(X_test)
print('Регрессия по K-ближайшим соседям:')
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(classification_report(y_test, y_pred))

Регрессия по K-ближайшим соседям:
Accuracy: 0.7134831460674157
              precision    recall  f1-score   support

       False       0.73      0.67      0.70        89
        True       0.70      0.75      0.72        89

    accuracy                           0.71       178
   macro avg       0.71      0.71      0.71       178
weighted avg       0.71      0.71      0.71       178



In [53]:
model = catboost = CatBoostClassifier(
    iterations=100,
    verbose=0,
    random_state=42
)

# Обучение модели
model.fit(X_train, y_train, eval_set=(X_test, y_test))

# Предсказание
y_pred = model.predict(X_test)
print('Регрессия CatBoost:')
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(classification_report(y_test, y_pred))

Регрессия CatBoost:
Accuracy: 0.651685393258427
              precision    recall  f1-score   support

       False       0.66      0.62      0.64        89
        True       0.64      0.69      0.66        89

    accuracy                           0.65       178
   macro avg       0.65      0.65      0.65       178
weighted avg       0.65      0.65      0.65       178



<b>Для лучшей модели попробуем автоматический подбор гипер параметров</b>

In [56]:
param_dist = {
    'n_neighbors': randint(1, 15),
    'weights': ['uniform', 'distance'],
    'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
    'leaf_size': randint(1, 100),
}
random_search = RandomizedSearchCV(
    estimator= KNeighborsClassifier(),
    param_distributions=param_dist,
    n_iter=500,
    scoring='roc_auc',
    cv=5,
    random_state=42,
    n_jobs=-10
)

random_search.fit(X_train, y_train)

best_params = random_search.best_params_
print("Лучшие параметры:", best_params)

best_model = random_search.best_estimator_
y_pred = best_model.predict(X_test)

print('Лучшие метрики Random Forest при подборе:')
print(f'Accuracy: {accuracy_score(y_test, y_pred)}')
print(classification_report(y_test, y_pred))


Лучшие параметры: {'algorithm': 'kd_tree', 'leaf_size': 15, 'n_neighbors': 11, 'weights': 'uniform'}
Лучшие метрики Random Forest при подборе:
Accuracy: 0.6741573033707865
              precision    recall  f1-score   support

       False       0.69      0.64      0.66        89
        True       0.66      0.71      0.68        89

    accuracy                           0.67       178
   macro avg       0.67      0.67      0.67       178
weighted avg       0.67      0.67      0.67       178



/home/simon/anaconda3/lib/python3.12/site-packages/numpy/ma/core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


<h2>Выводы</h2>

Для кластеризации препаратов по медианному значению <b>SI</b>, были использованы основные модели классического ML, такие как:

* LogisticRegression;
* DecisionTree;
* Random Forest;
* SVС;
* KNN.

А так же готовое решение CatBoostClassifier.

Наилучшую метрику Accuracy показала модель KNN со значением 0.71, что считается средним результатом. При этом метрика Recall составляет 0.67 и 0.75 для препаратов со значением ниже и выше медианного SI соответственно.

Для подбора наилучших гипер параметров был использован алгоритм случайного поиска. Но ручной подбор гипер параметров оказался лучше.